### DISCUSSION TITLE
Role of each UNet block

### DISCUSSION THREAD

**Discussion by cubiq** [cubiq/prompt_injection#1] [General]
I'm currently experimenting with SDXL and the nodes are dedicated to SDXL but this technique also works with SD1.5 and I'll release more nodes for that too but I believe SDXL has a stronger role for each block so I'd start with that.

So far what I've found is that **Input_8** and **Output_0** are mostly composition and **Output_1** is clearly only the style (you can try a prompt like `in the style of Van Gogh`).

**Input_7** has also a strong influence on the generation but I'm sure yet what it does exactly. **middle_0** also helps a little with overall composition.

The others blocks have to be decoded yet.

The following is a quick workflow to get you started: [sdxl_blocks.json](https://github.com/user-attachments/files/15733163/sdxl_blocks.json)

**Comment by Lalimec** [comment]
Here is a weird discovery. output_2 seems to have some sort of effect on the texts if there is any on the generated image.

![workflow (49)](https://github.com/cubiq/prompt_injection/assets/65521426/f2b08d26-5c9c-4880-8f9c-a7fd45dc5dbf)
![workflow (50)](https://github.com/cubiq/prompt_injection/assets/65521426/95961915-884c-45bc-8dde-c51f3feca59e)
![workflow (57)](https://github.com/cubiq/prompt_injection/assets/65521426/42874222-f6a4-4820-8fed-bd384eb10192)
![workflow (56)](https://github.com/cubiq/prompt_injection/assets/65521426/61206655-973e-41f3-88d2-70e7289fdeeb)
![workflow (55)](https://github.com/cubiq/prompt_injection/assets/65521426/f302a990-124e-4d75-9362-305100b6f0b4)
![workflow (54)](https://github.com/cubiq/prompt_injection/assets/65521426/e0d5ac24-ea05-4f8c-96d8-91d48ba5ae44)
![workflow (53)](https://github.com/cubiq/prompt_injection/assets/65521426/2b1924b8-927a-4d7b-b90c-5cb3d225f523)
![workflow (52)](https://github.com/cubiq/prompt_injection/assets/65521426/12f7b4f3-f5d0-4219-85c4-5fe2a676dc8f)
![workflow (51)](https://github.com/cubiq/prompt_injection/assets/65521426/0b067599-187c-456b-8fd8-613a3b8fdfdc)

**Reply by Lalimec** [reply]
![ComfyUI_02990_](https://github.com/cubiq/prompt_injection/assets/65521426/7b6d765a-151e-45a9-8599-afd6381e4989)
![ComfyUI_02992_](https://github.com/cubiq/prompt_injection/assets/65521426/a4deb4f9-44b1-4432-897b-62b52224690f)
![ComfyUI_02995_](https://github.com/cubiq/prompt_injection/assets/65521426/2c89b049-9191-4c78-9ab2-0fd848d65a92)
![ComfyUI_02991_](https://github.com/cubiq/prompt_injection/assets/65521426/25e7e02e-9022-4efa-b181-2d1a9c2ddd55)
![ComfyUI_02994_](https://github.com/cubiq/prompt_injection/assets/65521426/e6528d41-a14a-41a7-8a32-37cf55c2f3c4)
![ComfyUI_02993_](https://github.com/cubiq/prompt_injection/assets/65521426/2aa05ec0-fa59-4b25-b3f4-542ec78fa59e)

**Reply by cubiq** [reply]
this is interesting

**Reply by Lalimec** [reply]
what happens if we send ipadpter embeds to this block? like style trasnfer you mentioned in the video. is it possible to make some sort of "text transfer" i wonder.

**Reply by Lalimec** [reply]
some more tests https://imgur.com/a/SZ04M3e Seems like some of the text only effect the style, doesnt seem to write certain words.

**Comment by sdk401** [comment]
Input 7 looks like it is linked to color. But strangely, only for red. I've tried other colors and they do not work :)

this is the car with "red" in input_7
![image](https://github.com/cubiq/prompt_injection/assets/162196217/91ee25f0-ab8f-4a9d-933d-2a32033bc2e1)

and the same car with "red" in every other free input, but not in 7
![image](https://github.com/cubiq/prompt_injection/assets/162196217/e932e49c-3bc0-4e16-b04c-f1d18a677643)

Also, you can put "red" in all the other inputs and there will be no red in the picture. But in input 7 it's almost always present.

**Reply by Lalimec** [reply]
def does something lol 
![image](https://github.com/cubiq/prompt_injection/assets/65521426/4a751bf6-d798-45d3-957e-afd59ce4a191)
![image](https://github.com/cubiq/prompt_injection/assets/65521426/ad9353ca-cc3a-44c3-9efd-10c57bc5b84a)

**Comment by sdk401** [comment]
Also, the output_1 does not work exactly like style, more like "property". 

i9 seems to ignore any properties, but o1 reacts to them very strongly.

Examples:

"fresh, pink" in i9 and "old, dry" in o1
![image](https://github.com/cubiq/prompt_injection/assets/162196217/ee120da0-163f-4eee-b197-5960659b3ebe)

vice versa
![image](https://github.com/cubiq/prompt_injection/assets/162196217/67d7054c-21c2-4e5b-8e10-5222d9b28f34)

"old, dry" only in i9
![image](https://github.com/cubiq/prompt_injection/assets/162196217/66b79a7e-8b57-4842-8351-ede76407f728)

**Reply by Lalimec** [reply]
Yeah it's weird, gender is also tied to that i suppose.

![ss-240607195908](https://github.com/cubiq/prompt_injection/assets/65521426/f93a48c7-e3c7-43fa-899d-2c8184832fae)
![ss-240607195848](https://github.com/cubiq/prompt_injection/assets/65521426/c82546a0-2d7b-4639-96ef-05c140bada86)
![ss-240607200006](https://github.com/cubiq/prompt_injection/assets/65521426/ad1f1db1-5fa3-42af-9304-e3047610dcb8)
![ss-240607195931](https://github.com/cubiq/prompt_injection/assets/65521426/0932d0e7-1405-4560-8157-79be41b954b4)

**Comment by sdk401** [comment]
Ok, this is fun. Still a lot of bleeding, most of it happens in output_1, other fields ignore most of the properties.

![image](https://github.com/cubiq/prompt_injection/assets/162196217/07bd0646-58ac-4e65-aab2-a41345cb4673)

**Reply by sdk401** [reply]
Looks like putting the same condition in input_7 as in input_9 are reinforcing it a little.

with doubling:
![image](https://github.com/cubiq/prompt_injection/assets/162196217/a711ee63-5fb2-4054-a6e7-4be7d40ec032)

without:
![image](https://github.com/cubiq/prompt_injection/assets/162196217/8de74583-42d5-4685-bcc8-cddae6928f9b)

**Reply by sdk401** [reply]
Looks like inputs 7 and 0 are reinforcing actions, if they are mentioned in the main prompt. But not sure if that's the effect of just overall prompt reinforcing.
Also, time of the day and lighting is a property, and lives in output1, but camera angle is the main subject, and lives in output 0

**Comment by Lalimec** [comment]
Here is another test, we can see some of the tokens are not influencing the generation in each block. Some of the blocks doesnt even have any effect by them selves (including o2). It seems that some of them need certain conditions to be triggered. Also i7 really has a bias towards "red" lol.

![ss-240607202627](https://github.com/cubiq/prompt_injection/assets/65521426/3b59bd6c-2f17-4678-993a-cac3e3072a4b)
![ss-240607204725](https://github.com/cubiq/prompt_injection/assets/65521426/b2649e15-68f1-494c-aede-7be7f522cfdc)
![ss-240607202638](https://github.com/cubiq/prompt_injection/assets/65521426/ef1f0db7-c5c8-4891-a914-a729cb7a06d4)
![ss-240607202649](https://github.com/cubiq/prompt_injection/assets/65521426/ee305091-8e94-487b-b11f-60faeb3e3b19)
![ss-240607202708](https://github.com/cubiq/prompt_injection/assets/65521426/6442abf6-76fd-45db-bb7a-1a5d544ee38a)
![ss-240607202720](https://github.com/cubiq/prompt_injection/assets/65521426/39712a7f-f5d6-4611-aa3c-4efd38a33d80)
![ss-240607202731](https://github.com/cubiq/prompt_injection/assets/65521426/dfe65586-efeb-4794-b6e1-c6e06ee6b884)
![ss-240607202742](https://github.com/cubiq/prompt_injection/assets/65521426/ed4a5ec1-48a4-45a8-98b0-5165516bc723)
![ss-240607202753](https://github.com/cubiq/prompt_injection/assets/65521426/9d1b00c8-a7ab-44e3-894a-177cd77328f3)
![ss-240607202805](https://github.com/cubiq/prompt_injection/assets/65521426/cee60e07-6a9a-4fcf-9db6-eddcaf4d8f3b)
![ss-240607202815](https://github.com/cubiq/prompt_injection/assets/65521426/0ded8439-3824-4b12-ad47-5ab17c9aa33e)
![ss-240607202826](https://github.com/cubiq/prompt_injection/assets/65521426/069f0514-f9d9-4f02-9c01-856ac01ad543)
![ss-240607202837](https://github.com/cubiq/prompt_injection/assets/65521426/aa71b67b-bc2f-46a5-a045-3e0f52d9fb82)

**Comment by setothegreat** [comment]
Would like to add some input from my own basic testing:

Much like how you have to use different values on RAUnet and other Unet modifying nodes for 1.5 and XL, it seems like 1.5 models use their layers differently from XL.

In my experience:

Input 7 and 8 seem to work best for basic composition.
Output 4 seems to work best for more detailed composition.
Backgrounds seem to work best with Middle 0, but it also seems to drastically impact the rest of the image. Input 8 seems like it can affect the backgrounds slightly, but it's rather mild unless used in conjunction with Middle 0.
Haven't found any layers that respond well to style prompts, but that could just be the model I am using. Output 5 seems like it has the most effect, but even then it's rather subtle.

Seems like the individual Unet layers of 1.5 have more control over multiple aspects of generation than the XL models, meaning that prompting a combination of basic composition and background for Input 8 seems to produce better results overall.

Additionally, connecting the base prompt with Zero Out Conditioning to the positive prompt in the sampler seems to produce terrible results even if the base prompt is connected to the All input.

**Reply by cubiq** [reply]
I will release nodes specifically for SD1.5 at the moment I'm more concentrated on SDXL

**Reply by KINGLIFER** [reply]
> I will release nodes specifically for SD1.5 at the moment I'm more concentrated on SDXL
Cool!

**Reply by victorca25** [reply]
Thanks for the comment about the ZeroOut, I did not understand why it was so bad, but now I can continue testing with SD1.5 :D

For me, I found with the model I'm testing with:
- style to be controlled with output_3 and output_5 together
- facial expressions (smiling, angry, sad, etc) by output_4
- general background (mountain, beach, etc) by input_8

Will keep testing, but since multiple layers may have influence in multiple things, it might be required to concatenate conditions to have the proper conditioning in each layer

The original full prompt I'm using for testing is: "candid photo of a woman, high quality, detailed, smiling, natural skin" and here is an example of changing the above mentioned layers.

<img width="1199" alt="Screenshot 2024-06-10 at 19 55 23" src="https://github.com/cubiq/prompt_injection/assets/41912303/9a026a7c-35a2-4db0-8e53-a3d4867b55c5">

**Comment by crystalwizard** [comment]
i've got this set up at the moment 
![image](https://github.com/cubiq/prompt_injection/assets/17938346/f52dd27a-e59d-4006-b05b-ac1fefbec6ca)
to allow me to run the prompt into each block one at a time with no influence from anything else. It's proving very useful for figuring out what each block is doing. So far, most of them are ignoring the prompt entirely and just trying to refine the random noise. however input 8 paid attention to the prompt.

**Reply by crystalwizard** [reply]
[Unet-blocks-injection.json](https://github.com/user-attachments/files/15745225/Unet-blocks-injection.json)
here's my workflow if anyone wants to play around with it

**Comment by mpiquero1111** [comment]
just try to see of this has any influence with the original g and l 
![Screenshot 2024-06-07 195234](https://github.com/cubiq/prompt_injection/assets/28360938/c5298da1-fdad-4afb-a444-4fb4317e3d43)

**Reply by mpiquero1111** [reply]
quick test using what the others have found
![workflow (2)](https://github.com/cubiq/prompt_injection/assets/28360938/e18e8060-342c-437b-a803-10be0a330f70)

**Reply by mpiquero1111** [reply]
strong effect (affect?) 
![workflow (3)](https://github.com/cubiq/prompt_injection/assets/28360938/bb425017-d4dd-4c27-8d04-9725cf2a23d8)
didnt even have to use photoshop to do this so coooool

**Comment by crystalwizard** [comment]
after spending 12 hours on this today - input_8 is critical and seems to be the input that actually deals with the prompt itself. before that, the inputs are apparently working on the data from the model itself. 
plugging in to the bottleneck - middle - seems to have very little effect. I really wouldn't expect it to and don't think it's a good idea to inject data in there. 
output 5 seems to have the most effect of the outputs

**Reply by crystalwizard** [reply]
forgot to include that for all my tests, i've been using EulerA

**Comment by sdk401** [comment]
So far found two main "channels" (besides text in o2)

i8, i7, o0, main prompt:

- objects
- actions
- camera angle
- orientation (profile, frontal)
- size (stangely, not a property)

o1, descriptive properties:

- color
- age
- condition 
- time of the day
- time of the year
- lighting
- location flair (new york vs mumbai)
- weather, sky effects (stars, rain)

style and quality tags have effects on both channels, but "descriptive" channel effect is stronger.

to make some recognizable character you need to define him in both channels.

overall quality and coherence seems to improve when using "main" channel prompt on i9, i7 and o0, and putting "descriptive" prompt in everything else.

but testing together with regular prompt composed of same tensors i found no significant benefits of separating prompts to channels. bleeding seems around the same, and quality is not so different.

**Comment by JoeNavark** [comment]
My notes after a few hours of testing

Some schedulers seem to behave differently to prompt injection. Really broken prompts may occasionally result in broken images (pixelated, blurry) with normal that might work fine with AYS. Simple scheduler seems to always work. 

Core concepts are introduced in inputs 7(marginally), 8, and output 0 and 1. The effect of all other blocks are relative to how they relate to those concepts. Filling other blocks with a color, for example, may not induce any concept bleeding unless it relates to the core concept in some way. The bleeding also will tend to go away once it connects with the concept. A bunch of blue hair blocks may give you a blue shirt but once you have blue hair in a concept block the shirt won't be blue anymore.  

Attn2 unet does not cover the entire use of the prompt. Controlnet won't work just inside of attn2 but does do interesting and unpredictable things. 

I'd really like the ability to apply IPAdapter to specific blocks.

clipmultiplyattention can further increase the effects of prompts and make subtle changes more noticable.

**Reply by cubiq** [reply]
> I'd really like the ability to apply IPAdapter to specific blocks.

you can with the "mad scientist" node :smile:

**Comment by Regostratos** [comment]
Nice work, i have also found the same as you JoeNavark in relation to 7, 8, 0 and 1 being the main inputs. I also found Lalimec's discovery of the text on out 2 but I also found that if you add a strong input (Strength 5) to the out 5 it also improves the text consistency even if you use random input words. 

The other main difference is that i use conditioning average nodes into the inputs which means i can dial the strength either against a blank string input or against another prompt to give me the ability to tweak parts of the image till i get the result it like.
<img width="722" alt="Screenshot 2024-06-08 2214031" src="https://github.com/cubiq/prompt_injection/assets/36812531/a759d11a-0373-46dc-ba9b-56a4074aba3c"> Here is a screenshot of the text across multiple inputs. I used different colours on the Out 5 at a strength of 5 conditioning.

**Reply by cubiq** [reply]
remember that you can use weighted tokens `(word:1.5)` and you can also daisy chain the injection node to be able to use the strength and time stepping options

**Comment by Regostratos** [comment]
[
[Prompt Injection Testing v7.json.json](https://github.com/user-attachments/files/15748009/Prompt.Injection.Testing.v7.json.json)
](url)

Adjust the Conditioning strength to change change parts of the image

**Comment by tgitit** [comment]
**Emotional discovery**

Prompt injection is a fascinating project. Thankyou, Mateo!

I've been pondering for a while how to achieve more dynamic and consistent emotional dynamics in character expression. After some experimentation I think prompt injection may offer a significant step toward that.

Input8, middle0, output0, output1 and output2 appear to be where emotions live. 
An emotion (such as "joy", "amazement" etc) injected into various combinations of these parameters will produce greater or lesser degrees of emotional expression by a character, and sometimes bring other emotional variations.

Changing CFG also modulates the intensity of the emotional expression.

Inclusion of Input8 will sometimes give you a more dramatized version of the emotion, and occasionally include symbols.

Happy, joyful, gloating, jealous, envious, disbelieving, insulted, sympathetic, empathetic, caring, fascinated, upset, amazed, confused and unwell seem to be having an effect so far, and I expect there are many more. 

I've been zeroing out the pos and neg conditioning so only prompt injection was having an influence. 

However the emotional injection still appears to work well in combination with pos and neg prompts. 

Examples below also used ipadapter and faceid to transfer character likeness.


![1 confused](https://github.com/cubiq/prompt_injection/assets/134467613/fa150ea6-12fe-40cc-9095-0a3d9113e3c3)
Confused

![1a more confused](https://github.com/cubiq/prompt_injection/assets/134467613/9957ed9e-3fad-46a5-b32e-10cd6800cf76)
More confused

![2 amazed](https://github.com/cubiq/prompt_injection/assets/134467613/cc63b039-994f-4ae2-b6c4-5ae3ae0b219c)
Amazed

![3 happy](https://github.com/cubiq/prompt_injection/assets/134467613/6315a8aa-1ef5-462f-803b-e61f75654a2a)
Happy

![3a happier](https://github.com/cubiq/prompt_injection/assets/134467613/c456f683-50d1-4f2a-bbef-4f58e4f3273e)
Happy, joyful

**Reply by JoeNavark** [reply]
Since o2 has negligible impact on composition you can isolate it in it's own prompt injection and push the weight up a little higher to get more of that emotion while everything else in the image remains unchanged.

Unfortunately you can't go very high as o2 does start to distort the style of the image. Also it doesn't work for every emotion, only some are in o2 like angry.

**Comment by clayne** [comment]
Related prior work in a similar vein:

https://github.com/hnmr293/stable-diffusion-webui-dumpunet/blob/master/README_en.md
https://note.com/kohya_ss/n/n93b7c01b0547

**Comment by Creative-comfyUI** [comment]
The type of model will influence the way prompt injection will react. The model who support the fewer steps seems to give better results. Prompt injection doesn't work will all models.

**Comment by superprat** [comment]
Hey everyone, couple of observations on my experimentation with blocks - 
output_0 and output_1 have  the strongest effect on clothing. output_1 is the only block that has an effect on the color of the item.

A question I had was regarding the number of blocks available here. There are 12 in/out blocks and 1 middle on the UNET (do correct me if I'm wrong). Is there any reason all the blocks are not available for injection?

I am also experimenting with correlating block knowledge of concepts with lora training. If we are teaching specific concepts, perhaps we don't need to train all the blocks.

**Reply by Creative-comfyUI** [reply]
> Hey everyone, couple of observations on my experimentation with blocks - output_0 and output_1 have the strongest effect on clothing. output_1 is the only block that has an effect on the color of the item.
> 
> A question I had was regarding the number of blocks available here. There are 12 in/out blocks and 1 middle on the UNET (do correct me if I'm wrong). Is there any reason all the blocks are not available for injection?
> 
> I am also experimenting with correlating block knowledge of concepts with lora training. If we are teaching specific concepts, perhaps we don't need to train all the blocks.

I have discovered that model trained with few steps as turbo give the best results, the others give no results or bad results. 
It seems for the few steps all is concentrated on few blocks. 
Mixing a model with few steps with a model trained with more steps using Mergesdxl give the ability to block all blocks except 8 0 1 and in this case I obtain result that will not obtain if I mix it with MergeAdd or with the model with more steps. But the result is not  as good as the one with few steps. Look at my examples I have done on my page 
https://github.com/Creative-comfyUI/prompt_injection

**Comment by C1c3r0** [comment]
In my tests I was able to influence lighting in output_5, and can confirm that it helps with readability of the neon lights.

**Comment by toyxyz** [comment]
SD 1.5 and SDXL seem to have a slight difference in valid blocks. 
![image](https://github.com/cubiq/prompt_injection/assets/8006000/ba6b8b5f-cbc3-4ace-9c92-0ce89a642a4e)
For color/surface, output_5 shows the strongest influence. 
![image](https://github.com/cubiq/prompt_injection/assets/8006000/3a3842af-e9ee-4992-bfa7-4e1215fa4c19)
![image](https://github.com/cubiq/prompt_injection/assets/8006000/5bb180ce-3419-4188-85e1-a67a38b82316)
For scene compositions such as a number of people, input_7 and middle_0 look valid. 
![image](https://github.com/cubiq/prompt_injection/assets/8006000/6b1de8c0-3d0a-4eab-a796-43772a975d81)

**Reply by cubiq** [reply]
I added a new node that should be better compatible with SD1.5

**Reply by Creative-comfyUI** [reply]
Has this been done with mute prompt, zeroconditioning and only injection prompt? if it is the case will all have a prompt link to it? The sdxl model used, is it a kind of turbo sxdl (few steps)?

**Reply by victorca25** [reply]
@cubiq thanks for adding the new idx node including the missing SD1.5 blocks, just in case it's useful for others, the mapping from the standard node goes something like:
input4 -> idx2
input5 -> idx3
input7 -> idx4
input8 -> idx5
middle -> idx6
output3 -> idx7
output4 -> idx8
output5 -> idx9

output0, output1, output2 were not used for SD1.5, they have no cross attention layers

**Comment by proteus2000** [comment]
@cubiq thanks for your work. i am not sure how complicated the code would be, but a useful feature to add to your node would be the option to use a mask. this would allow you to more precisely select which part of the image you want to affect.

**Reply by cubiq** [reply]
yeah that is possible and in the cards

**Comment by Kinglord** [comment]
Hey all, if anyone is interested I did a deep dive on Pony UNet testing yesterday and posted my results and a basic WF here - #8 

Pony is a different beast so not sure if anything I've done there will be helpful but now you all have my current notes (subject to change of course lol).

**Comment by hologr4mSC** [comment]
[Repository of scene file and some images](https://github.com/hologr4mSC/unet-block-injection-testing/tree/main)

I've built a scene that automates prompt injection with easily identifiable characteristics that could help identify what each block does and what it influences. 
The scene also has some notes of my observations, hopefully it makes some sort of sense. 

the main scene file is here:
[unet_block_injection_automation_tests_001.json](https://github.com/hologr4mSC/unet-block-injection-testing/blob/main/unet_block_injection_automation_tests_001.json)

it works with any SDXL checkpoint, I don't know why the checkpoint I used is not on civitai anymore......

in the github project page are some example output images that can be loaded into ComfyUI to evaluate the inputs (don't run the scene, just load it) 

I'd appreciate any feedback

and if you want to see some eldritch horrors run a batch of outputs.

**Comment by setothegreat** [comment]
Been doing quite a bit of experimenting with this and thought I'd share my results.
Experiments have focused on prompting for the following aspects:

> **Subject:**
> The general overall look, composition and details of the subject. Nothing ultra specific, but more so the general ideas that would initially come to mind when you try to think of how the subject looks
> 
> **Composition:**
> The perspective, position, pose and such of the subject, or to put more simply, how a subject exists in the environment of the image
> 
> **Details:**
> More specific and fine details pertaining to the subject and composition; things that you wouldn't imagine when closing your eyes and imaging the subject or composition, but which help to give the subject and composition depth and believably
> 
> **Style:**
> How the image looks (photograph, anime, sketch, oil painting, etc)
> 
> **Background:**
> Anything pertaining to aspects which are not the immediate focus of the image, environmental or otherwise
> 
> **Text:**
> If the layer has an effect on the legibility of text

From this, I've developed the following chart: 

![image](https://github.com/cubiq/prompt_injection/assets/70491721/ad8e55d7-a661-4bf0-8446-7b44de5a5fb2)

The importance of the prompt type as it pertains to the image composition and impact on the layer have been weighted and order from left to right, meaning for instance that Input 8's prompt should focus more on composition than background prompts, and more on background prompts than style prompts.
Prompt types surrounded by brackets indicate that the effects of prompting can be positive or negative depending on the prompt, seed, and other variables, so prompt and use with caution.
Additionally, the color of the cell indicates whether you should use a separate text encoding node or not (since Output 2 to 5 don't have a huge impact on the image but do have a noticeable impact on the background, it usually makes sense to just use Input 8's node for simplicity).

I've also created the following note which goes into slightly more detail, with layers listed higher being more impactful in a positive way:

![image](https://github.com/cubiq/prompt_injection/assets/70491721/ae114b5f-7520-41f1-a660-06adff254ca5)

With all this being said, I have found that the best results seem to come from only having **Input 5, Input 7, Output 0, and Output 5** connected to their respective nodes, as having all the layers connected seems to really deepfry the image, though this conclusion could change with more testing depending on the specific use case.


Finally, 2 things which are probably of important to note: 


**First**, the base text encoder node connected to the "all" input should be a fairly all-encompassing description of the image you want without going overboard like a lot of prompts tend to. Try to keep it at around 50 tokens on average, and 80 tokens at max.

**And lastly**, alluding to the deep-frying I mentioned earlier, this method of prompt injection seems to really mess with the default CFG values you should use. 
Using my recommended Input 5, Input 7, Output 0, and Output 5 inputs, I've found that my images turn out far better in terms of both prompt adherence and image quality with a CFG value of 5, whereas without prompt injection a CFG value of 7.5 produces the best results for my specific model.
I would imagine further experiments with perturbed attention guidance could yield even better results, but I have yet to test them out. Will update when I can get around to it.

**Reply by setothegreat** [reply]
I'd additionally like to note that the weight value in the prompt injection node (at least the version I am using) doesn't seem to be functioning as intended:

How I would imagine the weight node ideally working is that it should weight the prompt as a ratio between the "all" and layer conditioning (meaning a weight of 0.8 would have a conditioning that is 80% the layer injected, and 20% the "all" node)

Instead though, the weight value seems to just impact the entirety of all the prompts input, including the "all" input. Since we are zeroing out the positive conditioning that's input into the KSampler, this seems to just result in the KSampler ignoring most of the prompt if you go below a weight of 80%, and usually results in images that do not adhere to anything but the negative prompt.

**Reply by cubiq** [reply]
the weight is added at the end to all embeds connected. 'ALL" is optional. The node can be daisy chained, the weight lets you calibrate the effect of all embeds connected to that one node.

**Reply by setothegreat** [reply]
> the weight is added at the end to all embeds connected. 'ALL" is optional. The node can be daisy chained, the weight lets you calibrate the effect of all embeds connected to that one node.

I think I understand what you're saying: so similarly to how the weight node works with your other extensions like IPAdapter, should we then be thinking of the Injector node as a way to modify how a model is conditioned, as opposed to a replacement for standard positive conditioning altogether?

If that's the case then that makes more sense, was just a little confused since the positive prompt was fully zeroed-out in the demonstration video

**Comment by xhoxye** [comment]
U-net block and attention layers Is there a relationship between them?
`Empirically, each layer of a deep network captures different semantic information, as we mentioned in Figure [7](https://arxiv.org/html/2404.02733v2#S3.F7), the key observation in our work is that there exists two specific attention layers handling style. Specifically, we find up_blocks.0.attentions.1 and down_blocks.2.attentions.1 capture style (color, material, atmosphere) and spatial layout (structure, composition) respectively. More intuitively, the 
4
𝑡
⁢
ℎ
 and the 
6
𝑡
⁢
ℎ
 as shown in. Figure [8](https://arxiv.org/html/2404.02733v2#S3.F8) for easy understanding. We can use them to implicitly extract style information, further preventing content leakage without losing the strength of the style. The idea is straightforward, as we have located style blocks, we can inject our image features into these blocks only to achieve style transfer seamlessly. Furthermore, since the number of parameters of the adapter is greatly reduced, the text control ability is also enhanced. This mechanism is applicable to other attention-based feature injection for editing or other tasks.`

**Reply by victorca25** [reply]
What do you mean? The cross attention layers are part of certain blocks in the UNet
